### PR TITLE
SIMD uint and ulong support

### PR DESCRIPTION
**Pull request by kbowers-jump**
... and minor comment cleanups.

This is a building block for future acceleration of various crypto algorithms in ballet.

### CODE REVIEW

File: src/util/simd/test_avx_4x64.c [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+    /* Reduction operations */
+
+    FD_TEST( !wc_any( wd_ne( wd_sum_all( x ), wd_bcast( x0 + x1 + x2 + x3 ) ) ) );
+    FD_TEST( !wc_any( wd_ne( wd_min_all( x ), wd_bcast( fmin( fmin( x0, x1 ), fmin( x2, x3 ) ) ) ) ) );
+    FD_TEST( !wc_any( wd_ne( wd_max_all( x ), wd_bcast( fmax( fmax( x0, x1 ), fmax( x2, x3 ) ) ) ) ) );
+
+    /* Misc operations */
+
+    wd_t m0; wd_t m1; wd_t m2; wd_t m3;
+    wd_transpose_4x4( wd_bcast( x0 ), wd_bcast( x1 ), wd_bcast( x2 ), wd_bcast( x3 ), m0, m1, m2, m3 );
```
Comment (ptaffet-jump) [inline src/util/simd/test_avx_4x64.c:158]: It might be nice to test transpose with something other than when all columns are identical

**Review by ptaffet-jump — APPROVED** [review]
Great! A few small comments, but it's fine to merge and address later

File: src/util/simd/fd_avx.h [hunk 2]
```diff
@@ -4,13 +4,14 @@
 #if FD_HAS_AVX
 
 /* An API for writing vectorized C/C++ code using 8-wide 32-bit ints,
-   8-wide 32-bit floats, 4-wide 64-bit doubles, 4-wide 64-bit longs and
-   8- or 4-wide logicals assuming a platform with AVX support.
+   8-wide 32-bit uints, 8-wide 32-bit floats, 4-wide 64-bit doubles,
+   4-wide 64-bit longs, 8-wide 64-bit ulongs and 8- or 4-wide logicals
```
[the remark this answers is not in the document]
Comment (kbowers-jump) [reply src/util/simd/fd_avx.h]: Done

File: src/util/simd/fd_avx_wi.h [hunk 3]
```diff
@@ -163,6 +163,22 @@ wi_insert_variable( wi_t a, int n, int v ) {
 #define wi_or(a,b)     _mm256_or_si256(     (a), (b) ) /* [   a0 |b0    a1 |b1 ...   a7 |b7 ] */
 #define wi_xor(a,b)    _mm256_xor_si256(    (a), (b) ) /* [   a0 ^b0    a1 ^b1 ...   a7 ^b7 ] */
 
+static inline wi_t wi_rol( wi_t a, int imm ) { return wi_or( wi_shl(  a, imm & 31 ), wi_shru( a, (-imm) & 31 ) ); }
+static inline wi_t wi_ror( wi_t a, int imm ) { return wi_or( wi_shru( a, imm & 31 ), wi_shl(  a, (-imm) & 31 ) ); }
+
+static inline wi_t wi_rol_variable( wi_t a, int n ) { return wi_or( wi_shl_variable(  a, n&31 ), wi_shru_variable( a, (-n)&31 ) ); }
+static inline wi_t wi_ror_variable( wi_t a, int n ) { return wi_or( wi_shru_variable( a, n&31 ), wi_shl_variable(  a, (-n)&31 ) ); }
```
[the remark this answers is not in the document]
Comment (kbowers-jump) [reply src/util/simd/fd_avx_wi.h:170]: Omitted this for the time being as it wasn't clear what the limiting principle was and the tests do provide very explicit documentation in the sense they are compared with (much more verbosely named) scalar bit ops like: `fd_uint_rotate_left`.  NBD if somebody wants to add these later and/or tweak the names (similarly for the shifts ... existing shift and rotates aren't too different from various asm opcode names in various instruction sets but aren't necessarily identical to any particular instruction set).

File: src/util/simd/fd_avx_wu.h [hunk 4]
```diff
@@ -0,0 +1,373 @@
+#ifndef HEADER_fd_src_util_simd_fd_avx_h
+#error "Do not include this directly; use fd_avx.h"
+#endif
+
+/* Vector uint API ****************************************************/
+
+/* A wu_t is a vector where each 32-bit wide lane holds an unsigned
+   32-bit twos-complement integer (a "uint").  These mirror wc and wf as
```
[the remark this answers is not in the document]
Comment (kbowers-jump) [reply src/util/simd/fd_avx_wu.h]: Done

File: src/util/simd/fd_avx_wu.h [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+  return _mm256_setr_epi32( i0, i1, i2, i3, i0, i1, i2, i3 );
+}
+
+static inline wu_t /* [ u0 u0 u1 u1 u2 u2 u3 u3 ] */
+wu_bcast_wide( uint u0, uint u1, uint u2, uint u3 ) {
+  int i0 = (int)u0; int i1 = (int)u1; int i2 = (int)u2; int i3 = (int)u3;
+  return _mm256_setr_epi32( i0, i0, i1, i1, i2, i2, i3, i3 );
+}
+
+/* No general vf_permute due to cross-128-bit lane limitations in AVX.
```
[the remark this answers is not in the document]
Comment (kbowers-jump) [reply src/util/simd/fd_avx_wu.h]: Done

File: src/util/simd/fd_avx_wu.h [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+static inline wu_t wu_ror_vector( wu_t a, wi_t b ) {
+  wi_t m = wi_bcast( 31 );
+  return wu_or( wu_shr_vector( a, wi_and( b, m ) ), wu_shl_vector( a, wi_and( wi_neg( b ), m ) ) );
+}
+
+static inline wu_t wu_bswap( wu_t a ) {
+  wu_t m = wu_bcast( 0x00FF00FFU );                                            /* Probably hoisted */
+  wu_t t = wu_rol( a, 16 );                                                    /* Swap E/O 16-bit pairs */
+  return wu_or( wu_andnot( m, wu_shl( t, 8 ) ), wu_and( m, wu_shr( t, 8 ) ) ); /* Swap E/O  8-bit pairs */
+}
```
[the remark this answers is not in the document]
Comment (kbowers-jump) [reply src/util/simd/fd_avx_wu.h:187]: No specific request here that I can see but am pleasantly surprised the compiler is going that deep (good sign the SIMD operation implementations on x86 are clean enough that the compiler can do deeper inferences and that the compiler is actually doing deeper inferences ... big step up from the days of yore).

The reason for the existing implementation is that a completely generalized byte oriented shuffle one of the glaring holes in the exposed historically functionality.  This intrinsic goes 80% of the way to fixing that (even better would be arbitrarily blend 2 vector registers into 1 ... that has the power to subsume the giant mess of very specialized one-off-y data swizzles on x86 into a single instruction that handle every conceivable data swizzle under the sun and has been done on non-Intel vector architectures for exactly this reason).

The existence of this intrinsic (assuming it is widely available in on community targets) hints that we probably generalize some of the 8-wide SIMD stuff for more general architectures.  But we can save that for another day.

File: src/util/simd/fd_avx_wu.h [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+     _mm256_cvtepi32_pd equivalent would not be exact such that add to
+     correct the twos complement mangling would add a possible second
+     roundoff error ... this would result in slightly different values
+     occassionally when u is >~ 2^31).  We instead convert the two
+     halves to double (exact), convert the double to float (single
+     roundoff error) and then concat the two float halves to make a
+     correctly rounded implementation. */
+
+  return _mm256_insertf128_ps( _mm256_insertf128_ps( _mm256_setzero_ps(), _mm256_cvtpd_ps( wu_to_wd( a, 0 ) ), 0 ),
+                                                                          _mm256_cvtpd_ps( wu_to_wd( a, 1 ) ), 1 );
```
[the remark this answers is not in the document]
Comment (kbowers-jump) [reply src/util/simd/fd_avx_wu.h]: Done.  (That's a lot nicer ... missed that in the easter egg hunt through the layers of petrified bit rot of Intel SIMD intrinsics.)

File: src/util/simd/fd_avx_wv.h [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+
+/* Given [a0 a1 a2 a3] and/or [b0 b1 b2 b3], return ... */
+
+/* Arithmetic operations */
+
+/* Note: _mm256_{min,max}_epu64 are missing pre AVX-512.  We emulate
+   these below.  Likewise, there is no _mm256_mullo_epi64 in AVX.  Since
+   this is not cheap to emulate, we do not provide a wv_mul for the time
+   being.  There is a 64L*64L->64 multiply (where the lower 32-bits will
+   be zero extended to 64-bits beforehand) though and that is very
```
[the remark this answers is not in the document]
Comment (kbowers-jump) [reply src/util/simd/fd_avx_wv.h]: Done

File: src/util/simd/fd_sse_vl.h [hunk 9]
```diff
@@ -177,14 +193,19 @@ static inline vl_t vl_shr_vector( vl_t a, vl_t n ) {
 
    vl_to_vc(d)     returns [ !!l0 !!l0 !!l1 !!l1 ] 
 
-   vl_to_vf(l,i,0) returns [ (float)l0 (float)l1 l2 l3 ]
+   vl_to_vf(l,i,0) returns [ (float)l0 (float)l1 f2 f3 ]
    vl_to_vf(l,i,1) returns [ f0 f1 (float)l0 (float)l1 ]
```
[the remark this answers is not in the document]
Comment (kbowers-jump) [reply src/util/simd/fd_sse_vl.h]: Done

**Review by ptaffet-jump — APPROVED** [review]

**Pull request merged**